### PR TITLE
KAFKA-4392: Handle NoSuchFileException gracefully in StateDirectory

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
This is just a quick fix

**Comment by enothereska** [comment]
LGTM. @dguy ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 1]
```diff
@@ -87,7 +88,16 @@ public boolean lock(final TaskId taskId, int retry) throws IOException {
...
         }
         final File lockFile = new File(directoryForTask(taskId), LOCK_FILE_NAME);
-        final FileChannel channel = getOrCreateFileChannel(taskId, lockFile.toPath());
+
+        final FileChannel channel;
+
+        try {
+            channel = getOrCreateFileChannel(taskId, lockFile.toPath());
+        } catch (NoSuchFileException e) {
+            // FileChannel.open(WRITE,APPEND) could throw NoSuchFileException during a startup and shutdown race
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: We seem to use `FileChannel.open(lockPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE)` (i.e. not APPEND)?

**Comment by guozhangwang** [comment]
@ijuma Updated comments.

**Comment by ijuma** [comment]
@dguy, can you please take a look?

**Comment by dguy** [comment]
sorry @ijuma and @guozhangwang - LGTM

**Comment by ijuma** [comment]
Thanks @dguy. I was wondering if it would be better to catch that exception in `getOrCreateFileChannel`?

**Comment by dguy** [comment]
And do what? Retry? It would be nice if there was a test, but it is a pretty tricky thing to write a test for.

**Comment by guozhangwang** [comment]
I agree It is a bit hard to write a unit test for it, and I can move the catch exception to `getOrCreateFileChannel` and return null, and let the caller function to return false when the returned value is null, but I feel it is a bit indirect in returning values. But if you have strong opinion let me know @ijuma .

**Comment by ijuma** [comment]
The method is only used from one place, so I don't feel strongly. I didn't understand the race condition though, what file is missing causing the NSFE?

**Comment by guozhangwang** [comment]
@ijuma Here is my analysis: with multiple threads, it is possible that when one thread is trying to create the lock file as `dir/appId/taskId/.lock`, the parent directory `dir/appId/taskId` has been deleted by another thread, causing the NSFE. One similar issue reported is this: https://bugs.openjdk.java.net/browse/JDK-8052403 

It has less relevance with the `CREATE, WRITE` file open options. I can update the comments and merge the patch if it sounds good to you.

**Comment by ijuma** [comment]
Sounds good to me.

**Comment by guozhangwang** [comment]
Previously I was confused because the exception was thrown on the lock file itself: 

```
java.nio.file.NoSuchFileException: /data/1/kafka-streams/myapp-streams/7_21/.lock
```

But looking at the above JDK ticket it seems to be an issue of `UnixFileSystemProvider.newFileChannel` itself that the exception message directly prints the local path even if it is actually the parent directory that is missing.

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Thanks, LGTM. Merged to trunk.

**Pull request closed without merging**